### PR TITLE
ECOM-2230: Added the catalog APIs.

### PR DESCRIPTION
**Pull request by aamir-khan**
WIP: Added the catalogs APIs.

@zubair-arbi @awais786 @tasawernawaz @ahsan-ul-haq Please review this PR.

### CODE REVIEW

File: ecommerce/extensions/api/v2/views.py [hunk 1]
```diff
@@ -550,7 +551,7 @@ def publish(self, request, pk=None):  # pylint: disable=unused-argument
 class ProductViewSet(NestedViewSetMixin, NonDestroyableModelViewSet):
     queryset = Product.objects.all()
     serializer_class = serializers.ProductSerializer
-    permission_classes = (IsAuthenticated, IsAdminUser,)
+    # permission_classes = (IsAuthenticated, IsAdminUser,)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views.py]: Delete commented-out code.

File: ecommerce/core/constants.py [hunk 2]
```diff
@@ -7,6 +7,9 @@
 COURSE_ID_REGEX = r'[^/+]+(/|\+)[^/+]+(/|\+)[^/]+'
 COURSE_ID_PATTERN = r'(?P<course_id>{})'.format(COURSE_ID_REGEX)
 
+# Regex used to match catalog IDs.
+CATALOG_ID_PATTERN = r'(?P<catalog_id>\d+)'
```
Comment (clintonb) [inline ecommerce/core/constants.py]: This is not needed.

**Comment by aamir-khan** [comment]
@clintonb @rlucioni Please review this PR.

File: ecommerce/extensions/api/serializers.py [hunk 3]
```diff
@@ -306,6 +306,18 @@ class Meta(object):
 
 
 class CatalogSerializer(serializers.ModelSerializer):
-    """Serializer for the Catalog object"""
+    """ Serializer for Catalogs. """
+
+    partner_id = serializers.IntegerField(source='partner.id')
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/serializers.py]: Can't you use `Partner` field already present in `Catalog` model?

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+from ecommerce.courses.models import Course
+from ecommerce.extensions.api.serializers import ProductSerializer
+from ecommerce.extensions.catalogue.tests.mixins import CourseCatalogTestMixin
+from ecommerce.tests.mixins import PartnerMixin, UserMixin
+
+
+Catalog = get_model('catalogue', 'Catalog')
+StockRecord = get_model('partner', 'StockRecord')
+Partner = get_model('partner', 'Partner')
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: This wouldn't be needed as you are using `PartnerMixin`.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        # Create partner and add course seat for that partner
+        self.edx_partner = self.create_partner('edx')
+        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
+        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner)
+
+        # Create stock record and add it to a catalog.
+        stock_record = StockRecord.objects.get(partner=self.partner)
+        self.catalog = Catalog.objects.create(name='dummy', partner=self.partner)
+        self.catalog.stock_records.add(stock_record)
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: Most of the setup in both classes seems similar. Why don't you make a `CatalogMixin` and move this into that. It would be helpful in the future as well.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+from ecommerce.courses.models import Course
+from ecommerce.extensions.api.serializers import ProductSerializer
+from ecommerce.extensions.catalogue.tests.mixins import CourseCatalogTestMixin
+from ecommerce.tests.mixins import PartnerMixin, UserMixin
+
+
+Catalog = get_model('catalogue', 'Catalog')
+StockRecord = get_model('partner', 'StockRecord')
+Partner = get_model('partner', 'Partner')
+Product = get_model('catalogue', 'Product')
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: This is not used.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        expires = datetime.datetime(2100, 1, 1, tzinfo=pytz.UTC)
+        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner, expires=expires)
+
+        # Create stock record and add it to a catalog.
+        self.catalog = Catalog.objects.create(name='dummy', partner=self.partner)
+        self.stock_record = StockRecord.objects.get(partner=self.partner)
+
+    def get_full_url(self, path):
+        """ Returns a complete URL with the given path. """
+        return 'http://testserver' + path
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: Use `get_full_url` implemented here https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/api/v2/tests/views/__init__.py#L46

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        response = self.client.get(path)
+        response_data = json.loads(response.content)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response_data['count'], 1)
+
+        expected_data = ProductSerializer(self.stock_record.product, context={'request': response.wsgi_request}).data
+        self.assertListEqual(
+            response_data['results'],
+            [expected_data]
+        )
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: you could do this in one line.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        # URL for getting catalog for partner.
+        self.url = reverse(
+            'api:v2:partner-catalogs-list',
+            kwargs={'parent_lookup_partner_id': self.edx_partner.id},
+        )
+
+    def get_full_url(self, path):
+        """ Returns a complete URL with the given path. """
+        return 'http://testserver' + path
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: same as above.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        data = {
+            'id': catalog.id,
+            'partner_id': catalog.partner.id,
+            'name': catalog.name,
+            'products': self.get_full_url(reverse(
+                'api:v2:catalog-product-list',
+                kwargs={'parent_lookup_stockrecords__stock_catalogs': catalog.id},
+            ))
+        }
+        return data
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: You are creating similar method `serialize_catalog` twice please move this in `CatalogMixin` as well.

File: ecommerce/extensions/api/v2/tests/views/test_catalogs.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        self.assertEqual(response.status_code, 200)
+        self.assertListEqual(json.loads(response.content)['results'], [expected_data])
+
+    def test_access_catalog_api(self):
+        """Verify only staff users can access the endpoint."""
+        self.client.logout()
+        user = self.create_user(is_staff=False)
+        self.client.login(username=user.username, password=self.password)
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 403)
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/test_catalogs.py]: check API access when user is not login as well.

File: ecommerce/extensions/api/serializers.py (hunk 3 above)
Comment (aamir-khan) [inline ecommerce/extensions/api/serializers.py]: `partner_id` this field was mention in docs not the `Partner`.

Comment (clintonb) [inline ecommerce/extensions/api/serializers.py]: `partner` is fine. There is no need for the extra field here.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 12]
```diff
@@ -1,81 +0,0 @@
-import json
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py:1]: Do the rename in a separate PR (to be merged first) so that we can see exactly what changed.

File: ecommerce/extensions/api/v2/views/catalog.py [hunk 13]
```diff
@@ -1,15 +1,15 @@
 from oscar.core.loading import get_model
+from rest_framework.viewsets import ReadOnlyModelViewSet
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/catalog.py:3]: Alphabetize.

File: ecommerce/extensions/catalogue/models.py [hunk 14]
```diff
@@ -19,7 +19,7 @@ class ProductAttributeValue(AbstractProductAttributeValue):
 class Catalog(models.Model):
     name = models.CharField(max_length=255)
     partner = models.ForeignKey('partner.Partner', related_name='catalogs')
-    stock_records = models.ManyToManyField('partner.StockRecord', blank=True)
+    stock_records = models.ManyToManyField('partner.StockRecord', blank=True, related_name='stock_catalogs')
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/models.py]: What's wrong with the default related name, `catalog_set`?

File: ecommerce/tests/mixins.py [hunk 15]
```diff
@@ -1,6 +1,8 @@
 # -*- coding: utf-8 -*-
 """Broadly-useful mixins for use in automated tests."""
 import json
+import datetime
+from datetime import timedelta
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: What's wrong with `datetime.timedelta`?

File: ecommerce/tests/mixins.py [hunk 16]
```diff
@@ -230,3 +237,39 @@ def create_partner(self, name):
         # Please make sure that 'name' is not longer than 8 characters.
         partner, __ = Partner.objects.get_or_create(short_code=name, name=name)
         return partner
+
+
+class CatalogMixin(PartnerMixin, CourseCatalogTestMixin, UserMixin, TestCase):
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: Test mixins should not inherit from `TestCase`.

File: ecommerce/tests/mixins.py [hunk 17]
```diff
@@ -230,3 +237,39 @@ def create_partner(self, name):
...
+        self.edx_partner = self.create_partner('edx')
+        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
+        expires = datetime.datetime.now() + timedelta(days=100)
+        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner, expires=expires)
+
+        # Create stock record and add it to a catalog.
+        self.catalog = Catalog.objects.create(name='dummy', partner=self.edx_partner)
+        self.stock_record = StockRecord.objects.get(partner=self.edx_partner)
+
+    def get_full_url(self, path):
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: Given that this is only used in one location, this doesn't need to be a method.

File: ecommerce/tests/mixins.py (hunk 16 above)
Comment (clintonb) [inline ecommerce/tests/mixins.py]: This mixin should reside closer to the tests that use it in `ecommerce/extensions/api/v2/tests/views/test_catalogs.py`.

File: ecommerce/tests/mixins.py [hunk 18]
```diff
@@ -230,3 +237,39 @@ def create_partner(self, name):
...
+class CatalogMixin(PartnerMixin, CourseCatalogTestMixin, UserMixin, TestCase):
+    """Provide methods for Catalog test cases."""
+
+    def setUp(self):
+        # Create and login the user with staff access.
+        self.user = self.create_user(is_staff=True)
+        self.client.login(username=self.user.username, password=self.password)
+
+        # Create partner and add course seat for that partner
+        self.edx_partner = self.create_partner('edx')
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: `CourseCatalogTestMixin` does this for you.

File: ecommerce/tests/mixins.py [hunk 19]
```diff
@@ -230,3 +237,39 @@ def create_partner(self, name):
...
+        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
+        expires = datetime.datetime.now() + timedelta(days=100)
+        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner, expires=expires)
+
+        # Create stock record and add it to a catalog.
+        self.catalog = Catalog.objects.create(name='dummy', partner=self.edx_partner)
+        self.stock_record = StockRecord.objects.get(partner=self.edx_partner)
```
Comment (clintonb) [inline ecommerce/tests/mixins.py]: Use `self.seat.stock_records.first()`.

File: ecommerce/extensions/api/v2/tests/views/mixins.py [hunk 20]
```diff
@@ -0,0 +1,42 @@
+import datetime
+from django.core.urlresolvers import reverse
+from oscar.core.loading import get_model
+
+from ecommerce.tests.mixins import PartnerMixin, UserMixin
+from ecommerce.courses.models import Course
+from ecommerce.extensions.catalogue.tests.mixins import CourseCatalogTestMixin
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/mixins.py:5]: Alphabetize

File: ecommerce/extensions/api/v2/tests/views/mixins.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+StockRecord = get_model('partner', 'StockRecord')
+
+
+class CatalogMixin(PartnerMixin, CourseCatalogTestMixin, UserMixin):
+    """Provide methods for Catalog test cases."""
+
+    def setUp(self):
+        super(CatalogMixin, self).setUp()
+        # Create and login the user with staff access.
+        self.user = self.create_user(is_staff=True)
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/mixins.py:19]: please change comment user is not logged in.

File: ecommerce/extensions/api/v2/tests/views/mixins.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+        # Create course seat for edx partner
+        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
+        expires = datetime.datetime.now() + datetime.timedelta(days=100)
+        self.seat = self.course.create_or_update_seat('honor', False, 0, self.partner, expires=expires)
+
+        # Create stock record and add it to a catalog.
+        self.catalog = Catalog.objects.create(name='dummy', partner=self.partner)
+        self.stock_record = StockRecord.objects.first()
```
Comment (ahsan-ul-haq) [inline ecommerce/extensions/api/v2/tests/views/mixins.py:27]: please change comment, no stock record is created here.

**Comment by aamir-khan** [comment]
@clintonb I have addressed your comments please review now.
@rlucioni  Please review this PR.

File: ecommerce/extensions/catalogue/models.py (hunk 14 above)
Comment (aamir-khan) [inline ecommerce/extensions/catalogue/models.py]: This was to make it more readable in related query.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 23]
```diff
@@ -4,52 +4,103 @@
...
+        response = self.client.get(self.catalog_list_path)
+        self.assertEqual(response.status_code, 403)
+
+    def test_catalog_list(self):
+        """Verify the endpoint returns all catalogs."""
+        response = self.client.get(self.catalog_list_path)
+        expected_data = self.serialize_catalog(self.catalog)
+        response_data = json.loads(response.content)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response_data['count'], 1)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py:49]: Validate the entire dictionary instead of the individual entries.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 24]
```diff
@@ -4,52 +4,103 @@
...
+        path = reverse('api:v2:catalog-detail', kwargs={'pk': self.catalog.id})
+        response = self.client.get(path)
+        self.assertEqual(response.status_code, 200)
+        self.assertDictEqual(json.loads(response.content), self.serialize_catalog(self.catalog))
 
-        # Create stock record and add it to a catalog.
-        stock_record = StockRecord.objects.get(partner=self.partner)
-        self.catalog = Catalog.objects.create(name='dummy', partner=self.partner)
-        self.catalog.stock_records.add(stock_record)
+    def test_catalog_create(self):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py]: This isn't necessary.

File: ecommerce/extensions/catalogue/models.py (hunk 14 above)
Comment (clintonb) [inline ecommerce/extensions/catalogue/models.py]: Use `catalogs`.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 25]
```diff
@@ -4,52 +4,103 @@
...
 
-        # Create and login the user with staff access.
-        self.user = self.create_user(is_staff=True)
         self.client.login(username=self.user.username, password=self.password)
 
-        # Create partner and add course seat for that partner
-        self.edx_partner = self.create_partner('edx')
-        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
-        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner)
+    catalog_list_path = reverse('api:v2:catalog-list')
```
Comment (rlucioni) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py]: Move this to the top of the class.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 26]
```diff
@@ -4,52 +4,103 @@
...
+
+    def test_access(self):
+        """Verify the only staff access."""
```
Comment (rlucioni) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py]: `"""Verify that only users with staff permissions can access the API."""`

File: ecommerce/extensions/api/v2/views/catalog.py [hunk 27]
```diff
@@ -1,15 +1,15 @@
 from oscar.core.loading import get_model
-from rest_framework_extensions.mixins import NestedViewSetMixin
 from rest_framework.permissions import IsAuthenticated, IsAdminUser
+from rest_framework.viewsets import ReadOnlyModelViewSet
 
+from rest_framework_extensions.mixins import NestedViewSetMixin
```
Comment (rlucioni) [inline ecommerce/extensions/api/v2/views/catalog.py:4]: This import belongs with the alphabetized group above.

**Comment by aamir-khan** [comment]
@edx/ecommerce Please review this PR.

File: ecommerce/extensions/api/v2/tests/views/mixins.py [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+    """Provide methods for Catalog test cases."""
+
+    def setUp(self):
+        super(CatalogMixin, self).setUp()
+        # Create the user with staff access.
+        self.user = self.create_user(is_staff=True)
+
+        # Create course seat for edx partner
+        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
+        expires = datetime.datetime.now() + datetime.timedelta(days=100)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/mixins.py]: Given that you aren't using the expiration datetime, you don't need to specify it.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 29]
```diff
@@ -4,52 +4,97 @@
...
 
-        # Create and login the user with staff access.
-        self.user = self.create_user(is_staff=True)
         self.client.login(username=self.user.username, password=self.password)
 
-        # Create partner and add course seat for that partner
-        self.edx_partner = self.create_partner('edx')
-        self.course = Course.objects.create(id='edX/DemoX/Demo_Course', name='Demo Course')
-        self.seat = self.course.create_or_update_seat('honor', False, 0, self.edx_partner)
+    def test_access(self):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py]: Break this out into two separate tests:
- `test_staff_authorization_required`
- `test_authentication_required`

File: ecommerce/extensions/catalogue/migrations/0011_auto_20151019_0639.py [hunk 30]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0011_auto_20151019_0639.py:1]: Where is the related model change for this?

Comment (rlucioni) [inline ecommerce/extensions/catalogue/migrations/0011_auto_20151019_0639.py:1]: @clintonb right below here, it would seem; `ecommerce/extensions/catalogue/models.py`.

**Comment by rlucioni** [comment]
:+1:

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py [hunk 31]
```diff
@@ -4,52 +4,97 @@
...
+        self.assertEqual(response.status_code, 403)
+
+    def test_catalog_list(self):
+        """Verify the endpoint returns all catalogs."""
+        response = self.client.get(self.catalog_list_path)
+        expected_data = self.serialize_catalog(self.catalog)
+        response_data = json.loads(response.content)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response_data['count'], 1)
+        self.assertListEqual(response_data['results'], [expected_data])
```
Comment (afeef) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py:50]: i believe @aamir-khan validated the dictionary here as suggested

**Comment by afeef** [comment]
@clintonb Please review

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py (hunk 23 above)
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py]: self.assertDictEqual(response_data, {'count': 1, 'results': [expected_data]})

**Comment by clintonb** [comment]
:+1:

**Comment by clintonb** [comment]
Remember to update your commit message. The fact that there is a squashed commit doesn't really matter.

File: ecommerce/extensions/api/v2/tests/views/test_catalog.py (hunk 23 above)
Comment (rlucioni) [inline ecommerce/extensions/api/v2/tests/views/test_catalog.py:49]: You can use `assertEqual` to perform this assertion; type-specific methods are automatically used by `assertEqual`.
